Possible connection issue in AbstractHttpDataPointProtobufReceiverConnection when entity.getContent().close() is called

Upon moving to SpringBoot 2.4.2, I’ve encountered what seems to be a connection issue with signalfx-java. SpringBoot 2.4.2 uses signalfx-java v1.0.5, but this may have been introduced with v1.0.4.

This can reproduced by modifying the test testHttpConnection as follows:
```
    public void testHttpConnection() throws Exception {
        Server server = new Server(0);
        server.setHandler(new AddPointsHandler());
        server.start();
        URI uri = server.getURI();
        DataPointReceiver dpr = new HttpDataPointProtobufReceiverFactory(
                new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort()))
                .createDataPointReceiver();
        dpr.addDataPoints(AUTH_TOKEN, Collections.singletonList(
                SignalFxProtocolBuffers.DataPoint.newBuilder().setSource("source").build()));
        
        Thread.sleep(30000);

        dpr.addDataPoints(AUTH_TOKEN, Collections.singletonList(
                SignalFxProtocolBuffers.DataPoint.newBuilder().setSource("source").build()));
        server.stop();
    }
```

Notice that it calls twice and sleeps 30 sec in between calls. Running the test results in in the following exception:

```
com.signalfx.metrics.SignalFxMetricsException: Exception posting to addDataPoints
	at com.signalfx.metrics.connection.AbstractHttpDataPointProtobufReceiverConnection.addDataPoints(AbstractHttpDataPointProtobufReceiverConnection.java:68)
	at com.signalfx.metrics.connection.HttpDataPointProtobufReceiverConnectionTest.testHttpConnection(HttpDataPointProtobufReceiverConnectionTest.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: org.apache.http.NoHttpResponseException: 192.168.0.12:51987 failed to respond
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:141)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at com.signalfx.connection.AbstractHttpReceiverConnection.postToEndpoint(AbstractHttpReceiverConnection.java:72)
	at com.signalfx.metrics.connection.AbstractHttpDataPointProtobufReceiverConnection.addDataPoints(AbstractHttpDataPointProtobufReceiverConnection.java:48)
	... 27 more
```

Debugging the issue took me into AbstractHttpDataPointProtobufReceiverConnection.java [here](https://github.com/signalfx/signalfx-java/blob/master/signalfx-java/src/main/java/com/signalfx/metrics/connection/AbstractHttpDataPointProtobufReceiverConnection.java#L61).  This change was introduced with signalfx-java v1.0.4 as part of this [PR](https://github.com/signalfx/signalfx-java/pull/91).  When I comment out `entity.getContent().close();`, I do not see the exception.

--- a/signalfx-java/src/main/java/com/signalfx/metrics/connection/AbstractHttpDataPointProtobufReceiverConnection.java
+++ b/signalfx-java/src/main/java/com/signalfx/metrics/connection/AbstractHttpDataPointProtobufReceiverConnection.java
@@ -21,6 +21,7 @@ import com.signalfx.connection.AbstractHttpReceiverConnection;
 import com.signalfx.endpoint.SignalFxReceiverEndpoint;
 import com.signalfx.metrics.SignalFxMetricsException;
 import com.signalfx.metrics.protobuf.SignalFxProtocolBuffers;
+import org.apache.http.util.EntityUtils;
 
 public abstract class AbstractHttpDataPointProtobufReceiverConnection extends AbstractHttpReceiverConnection implements DataPointReceiver {
 
@@ -56,11 +57,12 @@ public abstract class AbstractHttpDataPointProtobufReceiverConnection extends Ab
                 }
             } finally {
                 if (resp != null) {
-                    HttpEntity entity = resp.getEntity();
-                    if (entity != null) {
-                        entity.getContent().close();
+                    try {
+                        HttpEntity entity = resp.getEntity();
+                        EntityUtils.consume(entity);
+                    } finally {
+                        resp.close();
                     }
-                    resp.close();
                 }
             }
         } catch (IOException e) {

--- a/signalfx-java/src/main/java/com/signalfx/metrics/connection/HttpClientConnectionManagerFactory.java
+++ b/signalfx-java/src/main/java/com/signalfx/metrics/connection/HttpClientConnectionManagerFactory.java
@@ -10,7 +10,7 @@ import org.apache.http.conn.HttpClientConnectionManager;
 import org.apache.http.conn.socket.ConnectionSocketFactory;
 import org.apache.http.conn.socket.PlainConnectionSocketFactory;
 import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
-import org.apache.http.impl.conn.BasicHttpClientConnectionManager;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
 import org.apache.http.ssl.SSLContexts;
 
 public class HttpClientConnectionManagerFactory {
@@ -20,13 +20,13 @@ public class HttpClientConnectionManagerFactory {
   }
 
   public static HttpClientConnectionManager withTimeoutMs(int timeoutMs) {
-    BasicHttpClientConnectionManager httpClientConnectionManager = new BasicHttpClientConnectionManager(
+    PoolingHttpClientConnectionManager httpClientConnectionManager = new PoolingHttpClientConnectionManager(
         RegistryBuilder.<ConnectionSocketFactory>create()
             .register("http", PlainConnectionSocketFactory.getSocketFactory())
             .register("https", new SSLConnectionSocketFactoryWithTimeout(timeoutMs))
             .build());
 
-    httpClientConnectionManager.setSocketConfig(
+    httpClientConnectionManager.setDefaultSocketConfig(
         SocketConfig.custom().setSoTimeout(timeoutMs).build());
 
     return httpClientConnectionManager;

--- a/signalfx-java/src/test/java/com/signalfx/metrics/connection/HttpDataPointProtobufReceiverConnectionTest.java
+++ b/signalfx-java/src/test/java/com/signalfx/metrics/connection/HttpDataPointProtobufReceiverConnectionTest.java
@@ -3,6 +3,9 @@
  */
 package com.signalfx.metrics.connection;
 
+import com.signalfx.connection.AbstractHttpReceiverConnection;
+import com.signalfx.endpoint.SignalFxEndpoint;
+import com.signalfx.metrics.protobuf.SignalFxProtocolBuffers;
 import java.io.IOException;
 import java.net.URI;
 import java.nio.charset.StandardCharsets;
@@ -12,197 +15,221 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.zip.GZIPInputStream;
-
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
-
 import org.apache.http.HttpStatus;
 import org.apache.http.NameValuePair;
 import org.apache.http.client.utils.URLEncodedUtils;
 import org.apache.http.message.BasicNameValuePair;
+import org.eclipse.jetty.server.Connector;
 import org.eclipse.jetty.server.Request;
 import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
 import org.eclipse.jetty.server.handler.AbstractHandler;
 import org.junit.Test;
 
-import com.signalfx.connection.AbstractHttpReceiverConnection;
-import com.signalfx.endpoint.SignalFxEndpoint;
-import com.signalfx.metrics.protobuf.SignalFxProtocolBuffers;
-
 public class HttpDataPointProtobufReceiverConnectionTest {
 
-    public static final String AUTH_TOKEN = "AUTH_TOKEN";
+  public static final String AUTH_TOKEN = "AUTH_TOKEN";
 
-    @Test
-    public void testHttpConnection() throws Exception {
-        Server server = new Server(0);
-        server.setHandler(new AddPointsHandler());
-        server.start();
-        URI uri = server.getURI();
-        DataPointReceiver dpr = new HttpDataPointProtobufReceiverFactory(
-                new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort()))
-                .createDataPointReceiver();
-        dpr.addDataPoints(AUTH_TOKEN, Collections.singletonList(
-                SignalFxProtocolBuffers.DataPoint.newBuilder().setSource("source").build()));
-        server.stop();
+  @Test
+  public void testHttpConnection() throws Exception {
+    Server server = new Server(0);
+    server.setHandler(new AddPointsHandler());
+    server.start();
+    URI uri = server.getURI();
+    DataPointReceiver dpr = new HttpDataPointProtobufReceiverFactory(
+        new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort()))
+        .createDataPointReceiver();
+
+    dpr.addDataPoints(AUTH_TOKEN, Collections.singletonList(
+        SignalFxProtocolBuffers.DataPoint.newBuilder().setSource("source").build()));
+    server.stop();
+  }
+
+  @Test
+  public void shouldSendDataEvenIfServerClosedConnection() throws Exception {
+    int serverIdleTimeout = 5000;
+
+    Server server = new Server();
+    ServerConnector connector = new ServerConnector(server);
+    connector.setIdleTimeout(serverIdleTimeout);
+    connector.setPort(0);
+    server.setConnectors(new Connector[]{connector});
+    server.setHandler(new AddPointsHandler());
+    server.start();
+
+    URI uri = server.getURI();
+    DataPointReceiver dpr = new HttpDataPointProtobufReceiverFactory(
+        new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort()))
+        .createDataPointReceiver();
+
+    dpr.addDataPoints(AUTH_TOKEN, Collections.singletonList(
+        SignalFxProtocolBuffers.DataPoint.newBuilder().setSource("source").build()));
+
+    Thread.sleep(serverIdleTimeout + 1000);
+
+    dpr.addDataPoints(AUTH_TOKEN, Collections.singletonList(
+        SignalFxProtocolBuffers.DataPoint.newBuilder().setSource("source").build()));
+    server.stop();
+  }
+
+  @Test
+  public void testBackfill() throws Exception {
+    Server server = new Server(0);
+    server.setHandler(new BackfillHandler());
+    server.start();
+    URI uri = server.getURI();
+    DataPointReceiver dpr = new HttpDataPointProtobufReceiverFactory(
+        new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort()))
+        .createDataPointReceiver();
+
+    ArrayList<SignalFxProtocolBuffers.PointValue> values = new ArrayList<SignalFxProtocolBuffers.PointValue>(Arrays.asList(
+        SignalFxProtocolBuffers.PointValue.newBuilder().setTimestamp(System.currentTimeMillis())
+            .setValue(SignalFxProtocolBuffers.Datum.newBuilder().setDoubleValue(123.0)).build()
+    ));
+    HashMap<String, String> dims = new HashMap<String, String>();
+    dims.put("baz", "gorch");
+    dims.put("moo", "cow");
+
+    dpr.backfillDataPoints(AUTH_TOKEN, "foo.bar.baz", "counter", "ABC123", dims, values);
+    server.stop();
+  }
+
+  private class AddPointsHandler extends AbstractHandler {
+    @Override
+    public void handle(String target, Request baseRequest, HttpServletRequest request,
+                       HttpServletResponse response) throws IOException, ServletException {
+      if (!request.getHeader("X-SF-TOKEN").equals(AUTH_TOKEN)) {
+        error("Invalid auth token", response, baseRequest);
+        return;
+      }
+      if (!request.getHeader("User-Agent")
+          .equals(AbstractHttpReceiverConnection.USER_AGENT)) {
+        error("Invalid User agent: " + request.getHeader("User-Agent") + " vs "
+              + AbstractHttpReceiverConnection.USER_AGENT, response, baseRequest);
+        return;
+      }
+      SignalFxProtocolBuffers.DataPointUploadMessage all_datapoints =
+          SignalFxProtocolBuffers.DataPointUploadMessage.parseFrom(
+              new GZIPInputStream(baseRequest.getInputStream()));
+      if (!all_datapoints.getDatapoints(0).getSource().equals("source")) {
+        error("Invalid datapoint source", response, baseRequest);
+        return;
+      }
+      response.setStatus(HttpStatus.SC_OK);
+      response.getWriter().write("\"OK\"");
+      baseRequest.setHandled(true);
     }
 
-    @Test
-    public void testBackfill() throws Exception {
-        Server server = new Server(0);
-        server.setHandler(new BackfillHandler());
-        server.start();
-        URI uri = server.getURI();
-        DataPointReceiver dpr = new HttpDataPointProtobufReceiverFactory(
-                new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort()))
-                .createDataPointReceiver();
-
-        ArrayList<SignalFxProtocolBuffers.PointValue> values = new ArrayList<SignalFxProtocolBuffers.PointValue>(Arrays.asList(
-                SignalFxProtocolBuffers.PointValue.newBuilder().setTimestamp(System.currentTimeMillis())
-                        .setValue(SignalFxProtocolBuffers.Datum.newBuilder().setDoubleValue(123.0)).build()
-        ));
-        HashMap<String,String> dims = new HashMap<String,String>();
-        dims.put("baz", "gorch");
-        dims.put("moo", "cow");
-
-        dpr.backfillDataPoints(AUTH_TOKEN, "foo.bar.baz", "counter", "ABC123", dims, values);
-        server.stop();
+    private void error(String message, HttpServletResponse response, Request baseRequest)
+        throws IOException {
+      response.setStatus(HttpStatus.SC_BAD_REQUEST);
+      response.getWriter().write(message);
+      baseRequest.setHandled(true);
     }
 
-    private class AddPointsHandler extends AbstractHandler {
-        @Override
-        public void handle(String target, Request baseRequest, HttpServletRequest request,
-                           HttpServletResponse response) throws IOException, ServletException {
-            if (!request.getHeader("X-SF-TOKEN").equals(AUTH_TOKEN)) {
-                error("Invalid auth token", response, baseRequest);
-                return;
-            }
-            if (!request.getHeader("User-Agent")
-                    .equals(AbstractHttpReceiverConnection.USER_AGENT)) {
-                error("Invalid User agent: " + request.getHeader("User-Agent") + " vs "
-                        + AbstractHttpReceiverConnection.USER_AGENT, response, baseRequest);
-                return;
-            }
-            SignalFxProtocolBuffers.DataPointUploadMessage all_datapoints =
-                    SignalFxProtocolBuffers.DataPointUploadMessage.parseFrom(
-                            new GZIPInputStream(baseRequest.getInputStream()));
-            if (!all_datapoints.getDatapoints(0).getSource().equals("source")) {
-                error("Invalid datapoint source", response, baseRequest);
-                return;
-            }
-            response.setStatus(HttpStatus.SC_OK);
-            response.getWriter().write("\"OK\"");
-            baseRequest.setHandled(true);
-        }
-
-        private void error(String message, HttpServletResponse response, Request baseRequest)
-                throws IOException {
-            response.setStatus(HttpStatus.SC_BAD_REQUEST);
-            response.getWriter().write(message);
-            baseRequest.setHandled(true);
-        }
-
-        @Override
-        public boolean isRunning() {
-            return false;
-        }
-
-        @Override
-        public boolean isStarted() {
-            return false;
-        }
-
-        @Override
-        public boolean isStarting() {
-            return false;
-        }
-
-        @Override
-        public boolean isStopping() {
-            return false;
-        }
-
-        @Override
-        public boolean isStopped() {
-            return false;
-        }
-
-        @Override
-        public boolean isFailed() {
-            return false;
-        }
+    @Override
+    public boolean isRunning() {
+      return false;
     }
 
-    private class BackfillHandler extends AbstractHandler {
-        @Override
-        public void handle(String target, Request baseRequest, HttpServletRequest request,
-                           HttpServletResponse response) throws IOException, ServletException {
-            if(!request.getMethod().equals("POST")) {
-                error("Incorrect HTTP method for backfill", response, baseRequest);
-                return;
-            }
-            if(!request.getRequestURL().toString().endsWith("/v1/backfill")) {
-                error("Incorrect URL for backfill", response, baseRequest);
-            }
-
-            List<NameValuePair> params = URLEncodedUtils.parse(baseRequest.getQueryString(), StandardCharsets.UTF_8);
-            if(!params.contains(new BasicNameValuePair("orgid", "ABC123"))) {
-                error("orgid param is missing for backfill", response, baseRequest);
-            }
-            if(!params.contains(new BasicNameValuePair("metric_type", "counter"))) {
-                error("metric_type is missing for backfill", response, baseRequest);
-            }
-            if(!params.contains(new BasicNameValuePair("metric", "foo.bar.baz"))) {
-                error("metric is missing for backfill", response, baseRequest);
-            }
-            if(!params.contains(new BasicNameValuePair("sfdim_baz", "gorch"))) {
-                error("metric is missing for backfill", response, baseRequest);
-            }
-            if(!params.contains(new BasicNameValuePair("sfdim_moo", "cow"))) {
-                error("metric is missing for backfill", response, baseRequest);
-            }
-
-            response.setStatus(HttpStatus.SC_OK);
-            response.getWriter().write("\"OK\"");
-            baseRequest.setHandled(true);
-        }
-
-        private void error(String message, HttpServletResponse response, Request baseRequest)
-                throws IOException {
-            response.setStatus(HttpStatus.SC_BAD_REQUEST);
-            response.getWriter().write(message);
-            baseRequest.setHandled(true);
-        }
-
-        @Override
-        public boolean isRunning() {
-            return false;
-        }
-
-        @Override
-        public boolean isStarted() {
-            return false;
-        }
-
-        @Override
-        public boolean isStarting() {
-            return false;
-        }
-
-        @Override
-        public boolean isStopping() {
-            return false;
-        }
-
-        @Override
-        public boolean isStopped() {
-            return false;
-        }
-
-        @Override
-        public boolean isFailed() {
-            return false;
-        }
+    @Override
+    public boolean isStarted() {
+      return false;
     }
+
+    @Override
+    public boolean isStarting() {
+      return false;
+    }
+
+    @Override
+    public boolean isStopping() {
+      return false;
+    }
+
+    @Override
+    public boolean isStopped() {
+      return false;
+    }
+
+    @Override
+    public boolean isFailed() {
+      return false;
+    }
+  }
+
+  private class BackfillHandler extends AbstractHandler {
+    @Override
+    public void handle(String target, Request baseRequest, HttpServletRequest request,
+                       HttpServletResponse response) throws IOException, ServletException {
+      if (!request.getMethod().equals("POST")) {
+        error("Incorrect HTTP method for backfill", response, baseRequest);
+        return;
+      }
+      if (!request.getRequestURL().toString().endsWith("/v1/backfill")) {
+        error("Incorrect URL for backfill", response, baseRequest);
+      }
+
+      List<NameValuePair> params = URLEncodedUtils.parse(baseRequest.getQueryString(), StandardCharsets.UTF_8);
+      if (!params.contains(new BasicNameValuePair("orgid", "ABC123"))) {
+        error("orgid param is missing for backfill", response, baseRequest);
+      }
+      if (!params.contains(new BasicNameValuePair("metric_type", "counter"))) {
+        error("metric_type is missing for backfill", response, baseRequest);
+      }
+      if (!params.contains(new BasicNameValuePair("metric", "foo.bar.baz"))) {
+        error("metric is missing for backfill", response, baseRequest);
+      }
+      if (!params.contains(new BasicNameValuePair("sfdim_baz", "gorch"))) {
+        error("metric is missing for backfill", response, baseRequest);
+      }
+      if (!params.contains(new BasicNameValuePair("sfdim_moo", "cow"))) {
+        error("metric is missing for backfill", response, baseRequest);
+      }
+
+      response.setStatus(HttpStatus.SC_OK);
+      response.getWriter().write("\"OK\"");
+      baseRequest.setHandled(true);
+    }
+
+    private void error(String message, HttpServletResponse response, Request baseRequest)
+        throws IOException {
+      response.setStatus(HttpStatus.SC_BAD_REQUEST);
+      response.getWriter().write(message);
+      baseRequest.setHandled(true);
+    }
+
+    @Override
+    public boolean isRunning() {
+      return false;
+    }
+
+    @Override
+    public boolean isStarted() {
+      return false;
+    }
+
+    @Override
+    public boolean isStarting() {
+      return false;
+    }
+
+    @Override
+    public boolean isStopping() {
+      return false;
+    }
+
+    @Override
+    public boolean isStopped() {
+      return false;
+    }
+
+    @Override
+    public boolean isFailed() {
+      return false;
+    }
+  }
 }

